### ISSUE TITLE
$ref accepts all kind of arrays

### ISSUE THREAD

**Issue by fungiboletus** [xeipuuv/gojsonschema#270]
Hei,

The behaviour may be correct and accepted by the RFC but I find it weird. Using FHIR's json schema, `$ref` works as expected in extensions unless the user puts an array : all arrays are accepted. I guess it has something to do with the fact that extensions are arrays and using `$ref`.

In summary this is correctly refused:
```json
{
  "resourceType": "Patient",
  "extension": [{
    "url": "http://example.net/reference-test",
    "valueReference": { "wrongdocument": true }
  }]
}
```

And this one is not :
```json
{
  "resourceType": "Patient",
  "extension": [{
    "url": "http://example.net/reference-test",
    "valueReference": [
      { "wrongdocument": true }
    ]
  }]
}
```

I tried to make a summary of the JSON schema but [you can download the real one.](https://www.hl7.org/fhir/downloads.html)

```json
{
  "definitions": {
    "Reference": {
      "description": "blabla",
      "properties": {
        "reference": {
          "description": "blabla",
          "type": "string"
        }
      },
      "additionalProperties": false
    },
  "Patient": {
      "description": "blabla",
      "properties": {
        "resourceType": {
          "description": "blabla",
          "const": "Patient"
        },
        "extension": {
          "description": "blabla",
          "items": {
            "$ref": "#/definitions/Extension"
          },
          "type": "array"
        },
      "additionalProperties": false
    },
  "Extension": {
      "description": "blabla",
      "properties": {
        "url": {
          "description": "blabla",
          "type": "string",
        },
        "valueReference": {
          "description": "blabla",
          "$ref": "#/definitions/Reference"
        }
      },
      "additionalProperties": false
  }
}

**Comment by johandorland** [comment]
JSON Schema is not descriptive, but places constraints on a JSON instance. A `$ref` just applies the schema that it points to. If we look at the `"#/definitions/Reference"` schema:

<details>
<summary>Schema</summary>

```JSON
    "Reference": {
      "description": "A reference from one resource to another.",
      "properties": {
        "id": {
          "description": "Unique id for the element within a resource (for internal references). This may be any string value that does not contain spaces.",
          "$ref": "#/definitions/string"
        },
        "extension": {
          "description": "May be used to represent additional information that is not part of the basic definition of the element. To make the use of extensions safe and manageable, there is a strict set of governance  applied to the definition and use of extensions. Though any implementer can define an extension, there is a set of requirements that SHALL be met as part of the definition of the extension.",
          "items": {
            "$ref": "#/definitions/Extension"
          },
          "type": "array"
        },
        "reference": {
          "description": "A reference to a location at which the other resource is found. The reference may be a relative reference, in which case it is relative to the service base URL, or an absolute URL that resolves to the location where the resource is found. The reference may be version specific or not. If the reference is not to a FHIR RESTful server, then it should be assumed to be version specific. Internal fragment references (start with \u0027#\u0027) refer to contained resources.",
          "$ref": "#/definitions/string"
        },
        "_reference": {
          "description": "Extensions for reference",
          "$ref": "#/definitions/Element"
        },
        "type": {
          "description": "The expected type of the target of the reference. If both Reference.type and Reference.reference are populated and Reference.reference is a FHIR URL, both SHALL be consistent.\n\nThe type is the Canonical URL of Resource Definition that is the type this reference refers to. References are URLs that are relative to http://hl7.org/fhir/StructureDefinition/ e.g. \"Patient\" is a reference to http://hl7.org/fhir/StructureDefinition/Patient. Absolute URLs are only allowed for logical models (and can only be used in references in logical models, not resources).",
          "$ref": "#/definitions/uri"
        },
        "_type": {
          "description": "Extensions for type",
          "$ref": "#/definitions/Element"
        },
        "identifier": {
          "description": "An identifier for the target resource. This is used when there is no way to reference the other resource directly, either because the entity it represents is not available through a FHIR server, or because there is no way for the author of the resource to convert a known identifier to an actual location. There is no requirement that a Reference.identifier point to something that is actually exposed as a FHIR instance, but it SHALL point to a business concept that would be expected to be exposed as a FHIR instance, and that instance would need to be of a FHIR resource type allowed by the reference.",
          "$ref": "#/definitions/Identifier"
        },
        "display": {
          "description": "Plain text narrative that identifies the resource in addition to the resource reference.",
          "$ref": "#/definitions/string"
        },
        "_display": {
          "description": "Extensions for display",
          "$ref": "#/definitions/Element"
        }
      },
      "additionalProperties": false
    },
```
</details>

1: It describes the properties of the `"valueReference"` object _**IF**_ they are present
2: It says no additional properties are allowed

Nowhere does it say `"valueReference"` must be an object type. In these cases validation will just pass. JSON schema provides plenty of tools to restrict this further, but it requires a bit more effort on the schema author's side. You can find it weird that JSON schema is what it is, but I'm not going to create my own specification because a schema is too permissive.

TLDR; Blame the FHIR schema.

**Comment by fungiboletus** [comment]
Oh, so if I understand well, someone working for FHIR forgot to put `"type": "object"` everywhere ?

**Comment by johandorland** [comment]
Exactly, and in case some properties described in `"#/definitions/Reference"` are mandatory they should use `"required"` as well, to avoid an empty object being valid as well.

**Comment by fungiboletus** [comment]
On this specific `Reference` example, nothing is `"required"` and they have a text constraint : `At least one of reference, identifier and display SHALL be present (unless an extension is provided).` but I think it's out of JsonSchema's scope.

Thanks for the help.
